### PR TITLE
Implement guardian

### PR DESCRIPTION
**Pull request by PotterDai**
As is discussed here: https://github.com/AsynkronIT/protoactor-dotnet/issues/369

### CODE REVIEW

**Comment by PotterDai** [comment]
@rogeralsing 
OK, so this is what come up with:
https://github.com/PotterDai/protoactor-dotnet/blob/868a98295b21a6ea30241de5250c6d988dd57ccb/src/Proto.Actor/Guardians.cs

File: src/Proto.Cluster/PidCache.cs [hunk 1]
```diff
@@ -23,7 +23,7 @@ internal static class PidCache
 
         internal static void Setup()
         {
-            var props = Actor.FromProducer(() => new PidCacheWatcher());
+            var props = Actor.FromProducer(() => new PidCacheWatcher()).WithGuardianSupervisorStrategy(Supervision.KeepAliveStrategy);
```
Comment (rogeralsing) [inline src/Proto.Cluster/PidCache.cs]: IMO, AlwaysRestartStrategy would be a more descriptive name, as "KeepAlive" could mean anything, e.g. ignore errors etc.

Comment (PotterDai) [reply src/Proto.Cluster/PidCache.cs]: Yep, just got it updated.

**Comment by rogeralsing** [comment]
Looks great.

One thought, this PR uses the props style for setting guardian: `WithGuardianSupervisorStrategy`

That means that props containing this which are spawned in a child context.
e.g. `context.Spawn(props)` would have some strange semantics, or throw.
(Is the parent actor the supervisor, or is the guardian in this case?)

What if we have a copy of the Spawn methods on a Guardian class.

`var pid = RestartingGuardian.SpawnNamed("someActor",props);`

Not saying the above suggestion is the way to go, but we need to have a clear path for what happens to props used in a child context

**Comment by PotterDai** [comment]
Right, however considering there might be some cases where user want to use Guardian as well, RestartingGuardian.SpawnNamed will prohibit it from happening, for example when spawning in remote Activator.

I think we should just throw an exception to say that Guardian is not supported in context.Spawn ?
Or it might also be OK to just leave it as is (Maybe just print out an warning message), since GuardianStrategy will only be checked in Actor.Spawn.

**Comment by rogeralsing** [comment]
>I think we should just throw an exception to say that Guardian is not supported in context.Spawn ?

I'd say we should go with that. seems least surprising

**Comment by PotterDai** [comment]
Yep, just got it updated. Please check again = )

**Comment by rogeralsing** [comment]
Looks good!

**Comment by PotterDai** [comment]
I'll find a time to get this in Go as well = )

**Pull request merged**